File: Data Structures & Algorithms/binary-search/submission-0.java
class Solution {
    public int search(int[] nums, int target) {
        int lower = 0;
        int upper = nums.length - 1;
        

        while (lower <= upper) {
            int index = lower + (upper - lower) / 2;

            if (target > nums[index]) {
                lower = index + 1;
            } else if (target < nums[index]) {
                upper = index - 1;
            } else {
                return index;
            }
        }

        return -1;
    }
}

File: Data Structures & Algorithms/lowest-common-ancestor-in-binary-search-tree/submission-0.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right

class Solution:
    def lowestCommonAncestor(self, root: TreeNode, p: TreeNode, q: TreeNode) -> TreeNode:
        # possible answers: [p: q] inclusive
        lower = p.val
        upper = q.val
        if p.val > q.val:
            lower = q.val
            upper = p.val
        node = root

        while True:
            if node.val == lower or node.val == upper:
                return node
            elif lower < node.val < upper:
                return node
            elif node.val > upper:
                node = node.left
            elif node.val < lower:
                node = node.right


